src/misc/cloudAuth.ts

```diff
@@ -0,0 +1,194 @@
+        await this.initialize();
+
+        const item: ICloudAuthStorage = this.config.get('rcc');
+        if (new Date() < new Date(item.expiresAt)) {
+            return item.token.access_token;
+        }
+
+        await this.refreshToken();
+
+        return this.config.get('rcc.token.access_token', '') as string;
+    }
+
+    private async fetchToken(code: string | Array<string>): Promise<ICloudToken> {
+        try {
+            const request = {
+                grant_type: 'authorization_code',
+                redirect_uri: this.redirectUri,
+                client_id: clientId,
+                code,
+                code_verifier: this.codeVerifier,
+            };
+
+            const res = await axios.post(`${ cloudUrl }/api/oauth/token`, stringify(request));
+            const tokenInfo: ICloudToken = res.data;
+
+            const expiresAt = new Date();
+            expiresAt.setSeconds(expiresAt.getSeconds() + tokenInfo.expires_in);
+
+            const storageItem: ICloudAuthStorage = {
+                token: tokenInfo,
+                expiresAt,
+            };
+
+            this.config.set('rcc', storageItem);
+
+            return tokenInfo;
+        } catch (err) {
+            // tslint:disable-next-line:no-console
+            console.log(`[${ err.res.statusCode }] error getting token: ${ err.data.error } (${ err.data.requestId })`);
+
+            throw err;
+        }
+    }
+
+    private async refreshToken(): Promise<void> {
+        const refreshToken = this.config.get('rcc.token.refresh_token', '');
+
+        const request = {
+            client_id: clientId,
+            refresh_token: refreshToken,
+            scope,
+            grant_type: 'refresh_token',
+            redirect_uri: this.redirectUri,
+        };
+
+        try {
+            const res = await axios.post(`${ cloudUrl }/api/oauth/token`, stringify(request));
+            const tokenInfo: ICloudToken = res.data;
+
+            const expiresAt = new Date();
+            expiresAt.setSeconds(expiresAt.getSeconds() + tokenInfo.expires_in);
+
+            this.config.set('rcc.token.access_token', tokenInfo.access_token);
+            this.config.set('rcc.token.expires_in', tokenInfo.expires_in);
+            this.config.set('rcc.token.scope', tokenInfo.scope);
+            this.config.set('rcc.token.token_type', tokenInfo.token_type);
+            this.config.set('rcc.expiresAt', expiresAt);
+        } catch (err) {
+            const d = err.data;
+            // tslint:disable-next-line:no-console
+            console.log(`[${ err.res.statusCode }] error getting token refreshed: ${ d.error } (${ d.requestId })`);
+
+            throw err;
+        }
+    }
+
+    private buildAuthorizeUrl(codeChallenge: string) {
+        const data = {
+            client_id: clientId,
+            response_type: 'code',
+            scope,
+            redirect_uri: this.redirectUri,
+            state: uuidv4(),
```

Comment:
I know we have the code challenge.  But maybe we should also verify state match in the callback as per spec?

Reply:
Is the challenge not good enough?

Reply:
Probably good enough.. but we could be better then enough if we wanted :grin: 

Its fine with out it